packages/account/src/account.ts

```diff
@@ -558,6 +569,87 @@ export class Account extends AbstractAccount implements WithAddress {
     return this.sendTransaction(request);
   }
 
+  async consolidateCoins(params: { assetId: string }): Promise<SubmitAllCallbackResponse> {
+    const { assetId } = params;
+
+    const { coins } = await this.getCoins(assetId);
+
+    if (coins.length <= 1) {
+      throw new FuelError(ErrorCode.NO_COINS_TO_CONSOLIDATE, 'No coins to consolidate.');
+    }
+    const baseAssetId = await this.provider.getBaseAssetId();
+    const isBaseAsset = baseAssetId === assetId;
+
+    let submitAll: SubmitAllCallback;
+
+    if (isBaseAsset) {
+      ({ submitAll } = await this.assembleBaseAssetConsolidationTxs(coins));
+    } else {
+      throw new Error('implement me.');
+    }
+
+    return submitAll();
+  }
+
+  async assembleBaseAssetConsolidationTxs(coins: Coin[]) {
+    const chainInfo = await this.provider.getChain();
+    const maxInputsNumber = chainInfo.consensusParameters.txParameters.maxInputs.toNumber();
+
+    let totalFeeCost = bn(0);
+    const txs: ScriptTransactionRequest[] = [];
+    const batchSize = Math.ceil(coins.length / maxInputsNumber);
+    const gasPrice = await this.provider.estimateGasPrice(10);
+
+    for (let i = 0; i < batchSize; i += 1) {
+      const batchStart = i * maxInputsNumber;
+      const batchEnd = (i + 1) * maxInputsNumber;
+      const batch = coins.slice(batchStart, batchEnd);
+
+      // There is no reason to consolidate just one coin
+      if (batch.length > 1) {
+        const request = new ScriptTransactionRequest({
+          script: '0x',
+        });
+
+        request.addResources(batch);
+
+        const minGas = request.calculateMinGas(chainInfo);
+
+        const fee = calculateGasFee({
+          gasPrice,
+          gas: minGas,
+          priceFactor: chainInfo.consensusParameters.feeParameters.gasPriceFactor,
+          tip: request.tip,
+        });
+
+        request.maxFee = fee;
+
+        totalFeeCost = totalFeeCost.add(fee);
+
+        txs.push(request);
+      }
+    }
+
+    const submitAll: SubmitAllCallback = async () => {
+      const txResponses: TransactionResult<TransactionType.Script>[] = [];
+      const errors: FuelError[] = [];
+
+      for (const tx of txs) {
+        try {
+          const submit = await this.sendTransaction(tx);
+          const response = await submit.waitForResult<TransactionType.Script>();
+          txResponses.push(response);
+        } catch (error) {
+          errors.push(error as FuelError);
+        }
+      }
+
+      return { txResponses, errors };
+    };
+
+    return { txs, totalFeeCost, submitAll };
+  }
+
```

Comment:
Just thinking out loud here - you could externalize this into a method.

Pseudo-code.

```suggestion
    const submitAll = this.prepareSubmitAll({ txs });

    return { txs, totalFeeCost, submitAll };
  }

  async prepareSubmitAll(params: {
    txs: ScriptTransactionRequest[],
  }): SubmitAllCallback {

    return async () => {

      const txResponses: TransactionResult<TransactionType.Script>[] = [];
      const errors: FuelError[] = [];

      for (const tx of txs) {
        try {
          const submit = await this.sendTransaction(tx);
          const response = await submit.waitForResult<TransactionType.Script>();
          txResponses.push(response);
        } catch (error) {
          errors.push(error as FuelError);
        }
      }

      return { txResponses, errors };
    }

  };
``  `

Reply:
And this is how AI would make it hybrid:

```ts
  async prepareSubmitAll(params: {
    txs: ScriptTransactionRequest[],
    /** Mode of execution, defaults to 'sequential' */
    mode?: 'sequential' | 'parallel';
  }): SubmitAllCallback {
    // Default to 'sequential' if mode is not provided
    const { txs, mode = 'sequential' } = params;

    return async () => {
      const txResponses: TransactionResult<TransactionType.Script>[] = [];
      const errors: FuelError[] = [];

      if (mode === 'sequential') {
        // Sequential execution (original logic)
        for (const tx of txs) {
          try {
            const submit = await this.sendTransaction(tx);
            const response = await submit.waitForResult<TransactionType.Script>();
            txResponses.push(response);
          } catch (error) {
            errors.push(error as FuelError);
          }
        }
      } else {
        // Parallel execution
        const results = await Promise.allSettled(
          txs.map(async (tx) => {
            // Chain the transaction sending and result waiting
            const submit = await this.sendTransaction(tx);
            return await submit.waitForResult<TransactionType.Script>();
          })
        );

        // Process results from Promise.allSettled
        results.forEach((result) => {
          if (result.status === 'fulfilled') {
            txResponses.push(result.value);
          } else {
            // Ensure the rejected reason is treated as FuelError
            errors.push(result.reason as FuelError);
          }
        });
      }

      return { txResponses, errors };
    }
  };
```

Reply:
@arboleya This makes a lot of sense

Reply:
https://github.com/FuelLabs/fuels-ts/pull/3836/commits/820fd6b23cb48649140afc34943d42b455acff1a